File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    vector<vector<int>> dir = {{0,1}, {1, 0}, {0, -1}, {-1, 0}};
    void bfs(int r, int c, vector<vector<char>>& grid){
        int n = grid.size();
        int m = grid[0].size();
        queue<vector<int>> q; q.push({r,c});
        while(!q.empty()){
            int curRow = q.front()[0];
            int curCol = q.front()[1];
            q.pop();
            for(auto d: dir){
                int adjr = d[0]+curRow;
                int adjc = d[1]+curCol;
                if(adjr >= 0 && adjr < n && adjc >= 0 && adjc < m){
                    if(grid[adjr][adjc] == '1'){
                        //marks visited
                        grid[adjr][adjc] = '0';
                        q.push({adjr, adjc});
                    }
                }
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        int n = grid.size();
        int m = grid[0].size();
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == '1'){
                    count++;
                    //marks as visited
                    grid[i][j] = '0';
                    bfs(i, j, grid);
                }
            }
        }

        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution(object):
    def vanishIsland(self, grid, row, col):
        if grid[row][col] == '1':
            grid[row][col] = '0'
            
            # Checking up
            if row - 1 >= 0:
                self.vanishIsland(grid, row - 1, col)
            # Checking down
            if row + 1 < len(grid):
                self.vanishIsland(grid, row + 1, col)
            # Checking left
            if col - 1 >= 0:
                self.vanishIsland(grid, row, col - 1)
            # Checking right
            if col + 1 < len(grid[0]):
                self.vanishIsland(grid, row, col + 1)

    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.vanishIsland(grid, i, j)
                    count += 1
        return count
        